packages/frontend/src/features/sqlRunner/components/WriteBackToDbtModal.tsx

```diff
@@ -54,22 +58,45 @@ export const WriteBackToDbtModal: FC<Props> = ({ opened, onClose }) => {
         health?.hasGithub &&
         project?.dbtConnection.type === DbtProjectType.GITHUB
     );
-    const projectDirectory = useMemo(
-        () =>
-            canWriteToDbtProject
-                ? getProjectDirectory(project?.dbtConnection)
-                : undefined,
-        [project?.dbtConnection, canWriteToDbtProject],
-    );
-    const basePathForDbtCustomView =
-        `${projectDirectory}models/lightdash`.replace(/\/{2,}/g, '/');
-    const filePathsForDbtCustomView = useMemo(
-        () => [
-            `${snakeCaseName(form.values.name || 'custom view')}.sql`,
-            `${snakeCaseName(form.values.name || 'custom view')}.yml`,
-        ],
-        [form.values.name],
-    );
+
+    useEffect(() => {
+        // Initial preview load
+        if (!opened) return;
+
+        if (!writePreview && columns && !isPreviewLoading && !isPreviewError) {
+            void getWritePreview({
+                projectUuid,
+                name: 'custom view',
+                sql,
+                columns,
+            });
+        }
+    }, [
+        opened,
+        isPreviewLoading,
+        isPreviewError,
+        writePreview,
+        projectUuid,
+        sql,
+        columns,
+        getWritePreview,
+    ]);
+
+    useEffect(() => {
+        // Reload preview on name change
+        if (!opened) return;
+        const debounceTimer = setTimeout(() => {
+            if (form.values.name && projectUuid && sql && columns) {
+                void getWritePreview({
+                    projectUuid,
+                    name: form.values.name,
+                    sql,
+                    columns,
+                });
+            }
+        }, 300);
+        return () => clearTimeout(debounceTimer);
+    }, [opened, form.values.name, projectUuid, sql, columns, getWritePreview]);
```

Comment:
This is really, cool, but we can do two things here: 
- have only 1 `useEffect`
- use `mantine/hooks` for the debounced value
- keep the mutation's previous data so the UI doesn't jump on load

```tsx
// NOTE: This keeps a mutation's previous data (similar to useQuery's `keepPreviousData`
 const [writePreviewData, setWritePreviewData] =
        useState<ApiGithubDbtWritePreview['results']>();

    const form = useForm<FormValues>({
        initialValues: {
            name: '',
        },
        validate: zodResolver(validationSchema),
    });
    const { mutateAsync: getWritePreview, isLoading: isPreviewLoading } =
        useGithubDbtWritePreview(form.values.name);
    const [debouncedName] = useDebouncedValue(form.values.name, 300);
```

then the unified `useEffect` will be this

```tsx
  useEffect(() => {
        if (!opened || !projectUuid || !sql || !columns) return;

        const loadPreview = async () => {
            const data = await getWritePreview({
                projectUuid,
                name: debouncedName || 'custom view',
                sql,
                columns,
            });
            setWritePreviewData(data);
        };

        void loadPreview();
    }, [opened, projectUuid, debouncedName, sql, columns, getWritePreview]);

```

then, after the `Badge` we could do something like

```tsx
</Badge>
                            {isPreviewLoading && <Loader size="xs" />}
```

Reply:
I tried with 1 useEffect only, but then it was making multiple requests because it was triggered on `isLoading`but I'll try this approach. Thanks for the suggestion